code/modules/events/radiation_leak.dm

```diff
@@ -0,0 +1,142 @@
+/datum/round_event_control/radiation_leak
+	name = "Radiation Leak"
+	description = "A radiation leak happens somewhere on the station, emanating radiation around a machine in the area. \
+		Engineering can stop the leak by using certain tools on it."
+	typepath = /datum/round_event/radiation_leak
+	weight = 15
+	max_occurrences = 2
+	category = EVENT_CATEGORY_ENGINEERING
+
+/datum/round_event/radiation_leak
+	start_when = 1 // 2 seconds in
+	announce_when = 10 // 20 seconds in
+	end_when = 150 // 300 seconds / ~5 minutes in
+
+	var/datum/weakref/picked_machine
+
+/datum/round_event/radiation_leak/setup()
+	// Pick a xeno spawn somewhere in the world.
+	// We will try to find a machine within a few turfs of it to start spewing rads from.
+	var/list/possible_locs = GLOB.generic_event_spawns.Copy()
+	while(length(possible_locs))
+		var/turf/chosen_loc = get_turf(pick_n_take(possible_locs))
+		for(var/obj/machinery/sick_device in range(3, chosen_loc))
+			// Skip undertiles
+			if(sick_device.IsObscured())
+				continue
+			// Skip invisible stuff
+			if(sick_device.invisibility || !sick_device.alpha || !sick_device.mouse_opacity)
+				continue
+			// Look for dense machinery. Basically stops stuff like wall mounts and pipes, silly ones.
+			// But keep in vents and scrubbers. I think it's funny if they start spitting out radiation
+			if(!sick_device.density && !istype(sick_device, /obj/machinery/atmospherics/components/unary))
+				continue
```

Comment:
you should do the cheapest/most common checks first
so density/pipes, invisible things, then obscured